tests/feinterface/step-12.cc

```diff
@@ -0,0 +1,580 @@
+           ExcDimensionMismatch(values.size(), points.size()));
+
+    for (unsigned int i = 0; i < values.size(); ++i)
+      {
+        if (points[i](0) < 0.5)
+          values[i] = 1.;
+        else
+          values[i] = 0.;
+      }
+  }
+
+
+  template <int dim>
+  Tensor<1, dim>
+  beta(const Point<dim> &p)
+  {
+    Assert(dim >= 2, ExcNotImplemented());
+
+    Point<dim> wind_field;
+    wind_field(0) = -p(1);
+    wind_field(1) = p(0);
+    wind_field /= wind_field.norm();
+
+    return wind_field;
+  }
+
+
+
+  template <int dim>
+  class AdvectionProblem
+  {
+  public:
+    AdvectionProblem();
+    void
+    run();
+
+  private:
+    void
+    setup_system();
+    void
+    assemble_system();
+    void
+    solve();
+    void
+    refine_grid();
+    void
+    output_results(const unsigned int cycle) const;
+
+    Triangulation<dim>   triangulation;
+    const MappingQ1<dim> mapping;
+
+    FE_DGQ<dim>     fe;
+    DoFHandler<dim> dof_handler;
+
+    AffineConstraints<> constraints;
+
+    SparsityPattern      sparsity_pattern;
+    SparseMatrix<double> system_matrix;
+
+    Vector<double> solution;
+    Vector<double> right_hand_side;
+  };
+
+
+  template <int dim>
+  AdvectionProblem<dim>::AdvectionProblem()
+    : mapping()
+    , fe(1)
+    , dof_handler(triangulation)
+  {}
+
+
+  template <int dim>
+  void
+  AdvectionProblem<dim>::setup_system()
+  {
+    dof_handler.distribute_dofs(fe);
+
+
+    DynamicSparsityPattern dsp(dof_handler.n_dofs());
+    DoFTools::make_flux_sparsity_pattern(dof_handler, dsp);
+    sparsity_pattern.copy_from(dsp);
+
+    system_matrix.reinit(sparsity_pattern);
+    solution.reinit(dof_handler.n_dofs());
+    right_hand_side.reinit(dof_handler.n_dofs());
+  }
+
+
+  template <int dim>
+  void
+  AdvectionProblem<dim>::assemble_system()
+  {
+    typedef decltype(dof_handler.begin_active()) Iterator;
+    BoundaryValues<dim>                          boundary_function;
+
+    auto cell_worker = [&](const Iterator &  cell,
+                           ScratchData<dim> &scratch_data,
+                           CopyData &        copy_data) {
+      const unsigned int n_dofs = scratch_data.fe_values.get_fe().dofs_per_cell;
+      copy_data.reinit(cell, n_dofs);
+      scratch_data.fe_values.reinit(cell);
+
+      const auto &q_points = scratch_data.fe_values.get_quadrature_points();
+
+      const FEValues<dim> &      fe_v = scratch_data.fe_values;
+      const std::vector<double> &JxW  = fe_v.get_JxW_values();
+
+      for (unsigned int point = 0; point < fe_v.n_quadrature_points; ++point)
+        {
+          auto beta_q = beta(q_points[point]);
+          for (unsigned int i = 0; i < n_dofs; ++i)
+            for (unsigned int j = 0; j < n_dofs; ++j)
+              {
+                copy_data.cell_matrix(i, j) +=
+                  -beta_q * fe_v.shape_grad(i, point) *
+                  fe_v.shape_value(j, point) * JxW[point];
+              }
+        }
+    };
+
+    /*
```

Comment:
If you don't need this, can you remove it, or can you expand with a comment what this does differently from the other boundary worker?

Reply:
I tried to use FEInterfaceValues instead of FEValues for the boundary, but it doesn't make a lot of sense. Code is removed.